### PR TITLE
iox-#358 Add PR template for code contributions

### LINKED ISSUE #358: PR template for code contributions
## Brief feature description

Since we get more and more contributions, we should have a PR template with a checklist to help the new contributors and also the reviewer

## Detailed information

It would be nice to have a PR template with a checklist for the author and reviewer with all the items which are necessary to successfully merge a PR like correct commit messages, linked issues, etc.

### PR DESCRIPTION
**Pull request by jpsamper2009**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer

- This PR's description is an example of the PR template in action 😉 

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

- Closes #358

### CODE REVIEW

**Comment by jpsamper2009** [comment]
@elBoberido Feel free to make any suggestions or take it over from here.

**Comment by elBoberido** [comment]
@jpsamper2009 thank's for the proposal, I'll invite the committer and we'll discuss if/what changes are needed for our workflow

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elfenpiff) [inline .github/PULL_REQUEST_TEMPLATE/development.md:20]: We always need at least two reviewers and sometimes even more when its a hot topic. Therefore, I would adapt this a little bit. What about adding the Checklist for the PR Reviewer multiple times so that the reviewer can enter its name (in ENTER NAME) and then work on the list. Something like:

## Checklist for the PR Reviewer - Kermit
- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes
- [ ] Approve

## Checklist for the PR Reviewer - Lemonhead
- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes
- [ ] Approve

## Checklist for the PR Reviewer - ENTER NAME
- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes
- [ ] Approve

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elfenpiff) [inline .github/PULL_REQUEST_TEMPLATE/development.md:12]: Here I would add the following points:

[ ] Code formatted with clang-format.
[ ] Code according to our coding style and naming conventions.
[ ] Code change verified by unit test.
[ ] Public Code changes are documented via doxygen

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:20]: I'm not in favor of this. Do we really need to have multiple reviewers to check things like the correct PR title?

I also think we do not need the Approve checkbox and then there is nothing which is related to a single reviewer. Maybe we can change the Approve checkbox to a note like `PR will only be approved after all the checkboxes from above are ticked`

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: I like the idea, but maybe we can find a wording which makes it obvious that they can be ticked it they don't apply, e.g. for a PR which only fixes typos

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [inline .github/PULL_REQUEST_TEMPLATE/development.md:26]: This probably should become `## Post-review Checklist for the Eclipse Committer` since not all contributors have the right to merge

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elfenpiff) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: Rephrase the third point to:
[ ] Code behavior change verified by unit test.

So typos aren't a problem anymore.

Comment (budrus) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: I'm not sure if one can rely verify that all lines have seen the clang-format. The box would be ticked if the reviewer had the feeling that it looks good? I feel that here more could be automatically checked

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (budrus) [reply .github/PULL_REQUEST_TEMPLATE/development.md:20]: I would propose that when one reviewer checked something and ticked the box, others do not have to do this explicitly again. The additional reviewers are more useful for the non-formal checks

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (budrus) [inline .github/PULL_REQUEST_TEMPLATE/development.md:24]: Approve is a bit redundant or?

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (budrus) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: The test for the changed behavior is a good thing

**Comment by elBoberido** [comment]
I'll wait till the discussions settles a bit, then I will update the PR on Monday after lunch with the non controversial changes

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: > Rephrase the third point to:
> [ ] Code behavior change verified by unit test.
> 
> So typos aren't a problem anymore.

How about `[ ] Code changes verified by unit tests where applicable`. Same about the doxygen point.

**Comment by dkroenke** [comment]
An additional check of the Post-review Checklist for the PR Author could be that all open points are addressed and tracked via issues

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: on the other side, item which are not applicable could be crossed out. We should maybe state that at the top of the template

**Comment by elBoberido** [comment]
I think we should also add a pre-review check for an update of the copyright owners

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (ithier) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: The third point is not clear to me. "Code change verified by unit test.".

Does this mean:
1. Unit tests are passing
2. Unit tests have been written for new behaviour ?

If it means the latter, I would change this point explicitly to "Unit tests have been written for new behaviour".

Comment (ithier) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: [ ] Code formatted with clang-format.

This should not be a review point.
We need to update the workflows to check it and it should fail when clang format has not been applied.
There must be a way to do this.

Comment (ithier) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: [ ] Public Code changes are documented via doxygen

This should be renamed to "Public APIs" ?
All code changes are public on github ;).

Comment (ithier) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: Maybe add "[ ] Relevant issues are linked" ?

Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:12]: > The third point is not clear to me. "Code change verified by unit test." (and also " Code changes verified by unit tests where applicable").
> 
> Does this mean:
> 
>     1. Unit tests are passing
> 
>     2. Unit tests have been written for new behaviour ?
> 
> 
> If it means the latter, I would change this point explicitly to "Unit tests have been written for new behaviour".

@ithier if they are not passing, the PR cannot be merged ;)

**Comment by elBoberido** [comment]
I've updated the PR description and the template.
Where I'm not quite sure are these items
```
1. [x] Code according to our coding style and naming conventions
1. [x] Unit tests have been written for new behavior
1. [x] Public API changes are documented via doxygen
1. [x] Copyright owner are updated in the changed files
```
Currently they are at the author checklist, but they could also fit the reviewer checklist

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (marthtz) [inline .github/PULL_REQUEST_TEMPLATE/development.md:23]: Do the notes also include a brief description of the changes? Or is mentioned somewhere else?

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (marthtz) [inline .github/PULL_REQUEST_TEMPLATE/development.md:3]: I just noticed that we don't explicitly mention that `#123` is an issue number. Should we add that to CONTRIBUTING.md? So an issue has to be created prior to creating a branch? Just remembered as there was some confusion in PR #345 about this.

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:23]: yes, I would assume so

File: .github/PULL_REQUEST_TEMPLATE/development.md
Comment (elBoberido) [reply .github/PULL_REQUEST_TEMPLATE/development.md:3]: you are right. It's not yet in CONTRIBUTION.md. I will add it

**Comment by budrus** [comment]
> 
> 
> I've updated the PR description and the template.
> Where I'm not quite sure are these items
> 
> ```
> 1. [x] Code according to our coding style and naming conventions
> 1. [x] Unit tests have been written for new behavior
> 1. [x] Public API changes are documented via doxygen
> 1. [x] Copyright owner are updated in the changed files
> ```
> 
> Currently they are at the author checklist, but they could also fit the reviewer checklist

My first feeling is to move these to the reviewer checklist

**Review by ithier — DISMISSED** [review]

**Review by marthtz — DISMISSED** [review]

**Comment by elBoberido** [comment]
I think this settled now. If there aren't any objections, I would like to merge this in the evening

**Review by elBoberido — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]

**Review by ithier — APPROVED** [review]

**Pull request merged**
